pyro/poutine/subsample_messenger.py

```diff
@@ -149,4 +149,17 @@ def _postprocess_message(self, msg):
                             .format(self.name, self.size, self.dim, statement, shape))
                     # Subsample parameters with known batch semantics.
                     if self.subsample_size < self.size:
-                        msg["value"] = msg["value"].index_select(dim, self._indices)
+                        value = msg["value"]
```

Comment:
why was the following logic necessary?

Reply:
The logic is used to add `_pyro_subsample` attribute to the parameters which need to be optimized. I think that checking `grad=0` might affect non-subsampled parameters (e.g. when `loss = 2 * param +...` and `param` has the initial value is `0`).

Reply:
It seems cleanest to me to use purely mask-based logic if possible, and assume it "almost surely" works. Can you think of a counterexample other than the first step example above?

Reply:
I can't come up with a good example which shows the shortcomings of using `mask = (grad != 0)`: one example I have in mind is when a parameter moves to a local maximum area, which can have nonzero measure due to precision issue - to get out of it, we need to update the parameter by using grad information from past steps.